scalafmt-sbt/src/main/scala/org/scalafmt/sbt/ScalafmtPlugin.scala

```diff
@@ -1,25 +1,194 @@
+              // do nothing
+              None
+            } else {
+              Some(onError(file, e))
+            }
+          case Formatted.Success(code) =>
+            Some(onFormat(file, input, code))
+        }
+      }
+    )
+  }
+
+  private def formatSources(
+      sources: Seq[File],
+      config: ScalafmtConfig,
+      log: Logger
+  ): Unit = {
+    withFormattedSources(sources, config)(
+      (file, e) => log.error(s"Error in ${file.toString}: $e"),
+      (file, input, output) => {
+        if (input != output) {
+          log.info(s"Successfully formatted ${file.toString}.")
+          IO.write(file, output)
+        }
+      }
+    )
+    PlatformTokenizerCache.megaCache.clear()
+  }
+
+  private def checkSources(
+      sources: Seq[File],
+      config: ScalafmtConfig,
+      log: Logger
+  ): Boolean = {
+    val res = withFormattedSources(sources, config)(
+      (file, e) => {
+        log.error(s"Error in ${file.toString}: $e")
+        false
+      },
+      (file, input, output) => {
+        val diff = input != output
+        if (diff) {
+          throw new MessageOnlyException(
+            s"${file.toString} isn't formatted properly!")
+        }
+        diff
       }
+    ).flatten.forall(x => x)
+    PlatformTokenizerCache.megaCache.clear()
+    res
   }
-  override def globalSettings: Seq[Def.Setting[_]] =
-    Seq(
-      commands += autoImport.scalafmt
-    ) ++
-      addCommandAlias("scalafmtTest", "scalafmt --test") ++
-      addCommandAlias("scalafmtDiffTest", "scalafmt --diff --test") ++
-      addCommandAlias("scalafmtDiff", "scalafmt --diff")
 
+  private lazy val sbtSources = thisProject.map(
+    proj => {
+      val rootSbt =
+        BuildPaths.configurationSources(proj.base).filterNot(_.isHidden)
+      val projectSbt =
+        (BuildPaths.projectStandard(proj.base) * GlobFilter("*.sbt")).get
+          .filterNot(_.isHidden)
+      rootSbt ++ projectSbt
+    }
+  )
+  private lazy val projectSources = thisProject.map(proj =>
+    (BuildPaths.projectStandard(proj.base) * GlobFilter("*.scala")).get)
+
+  lazy val scalafmtConfigSettings: Seq[Def.Setting[_]] = Seq(
+    scalafmt := formatSources(
+      (unmanagedSources in scalafmt).value,
+      scalaConfig.value,
+      streams.value.log),
+    scalafmtSbt := {
+      formatSources(sbtSources.value, sbtConfig.value, streams.value.log)
+      formatSources(projectSources.value, scalaConfig.value, streams.value.log)
+    },
+    scalafmtCheck :=
+      checkSources(
+        (unmanagedSources in scalafmt).value,
+        scalaConfig.value,
+        streams.value.log),
+    scalafmtSbtCheck := {
+      checkSources(sbtSources.value, sbtConfig.value, streams.value.log)
+      checkSources(projectSources.value, scalaConfig.value, streams.value.log)
+    },
+    scalafmtDoFormatOnCompile := Def.settingDyn {
+      if (scalafmtOnCompile.value) {
+        scalafmt in resolvedScoped.value.scope
```

Comment:
This will run scalafmt on **all** sources on **every** `compile`, including nop compiles. It does not take advantage of incrementality. Is this expected/intended behavior? cc/ @fommil I think scalariform uses FileFunction.cached to make this incremental https://github.com/sbt/sbt-scalariform/blob/a378b724a76ee94819939efc0e9adebd896e57d9/src/main/scala/com/typesafe/sbt/Scalariform.scala#L85

Reply:
yeah it's only supposed to be compiling files that are being compiled, not *all* of them

Reply:
Let's leave caching for next PRs. I would like to implement it entirely, not only in some specific places. I will write a warning in the task description for now :)

Reply:
@olafurpg that's a wrong solution. I don't know why scalariform uses it. `FileFunction.cached` accepts set of files, if you modify them (and we modify them by formatting), then it thinks that files are changed and runs the function again. So it needs two iteration to "converge" to formatted source and forget about it.